design-documents/template-versioning.md

```diff
@@ -0,0 +1,87 @@
+## Magento Frontend Template Versioning
+
+### Terms
+
+<!-- Describe any new terms used in the document -->
+
+### Overview
+
+Let's say ModuleA has template foo.phtml and CoolTheme overrides foo.phtml. Magento makes changes to foo.phtml and other files in ModuleA. Now CoolTheme has several problems:
+
+- Hard to know if there were changes to templates it has replaced 
+- Hard to know if how essential the changes are (security fix, new feature, typo, etc)
+- Hard to know which files to even update 
+- Hard to know which of the updates need to be applied to their overrides from the changes made to the base module template 
+
+To make this problem even worse there have no rules surrounding backwards compatibility of templates in general. The DOM structure can change without notice and our automated tests will pass and all extensions will break with all of the above problems when trying to fix them without any rule violations.
+
+These are very common scenarios and they have large security implications as well. Let's say there are some security fixes that spans across many templates. This could be the simple addition of a new hidden field, or something else like a new escapement policy. Whether it affects one or all of our template files, only a select few will actually ever be updated on merchant websites due to overrides and the difficulty of upgrading the templates.
+
+### Design
+
+
+At a very high level this proposal aims to introduce several new things:
+
+- A template versioning system with backwards compatibility rules for version changes
+- Merchant visibility into theme compatibility
+
+#### Template versioning
+
+A system of rules for html and phtml files using a new `@version x.x.x` annotation that allows templates to be updated in a way that allows developers to not only answer the problems listed above but also to handle the updating of their overrides in a much more controlled and expected way. The rules for versioning would also aim to prevent developers from creating backwards incompatible template changes between magento versions.
```

Comment:
Likely a naive question - do we use the `@version x.x.x` annotations anywhere else, or would this be a new pattern for versioning?

Reply:
We don't. We use `@since` for some `@api` related code but that doesn't solve the problem we are trying to solve here. We could definitely use `@since` as well as part of the new rules.

Reply:
> We could definitely use @since as well as part of the new rules.

No idea what we use `@since` for so I won't comment there :)

Can you give an example (comment or in doc, either way) of what this would look like in a template with some modifications?

Reply:
`@since` indicated when a specific class, method, parameter, etc, has been added. It says that thing has been available `@since 100.0.2` for example. It doesn't solve the problem of changes discussed but we could annotate DOM features or output like WordPress does. We still need something to indicate when and how significant the template was modified.

Reply:
> we could annotate DOM features or output like WordPress does

Would _love_ to see any references to that in this document as you flesh out ideas more, just for the sake of prior art research for the rest of us reviewing.

I have never touched WP so I'm ignorant here :)

Reply:
Sure. I'll update the doc in the near future but for now, here's an example of a woocommerce template with both a version in the template itself as well as annotated features for when they were added https://github.com/woocommerce/woocommerce/blob/master/templates/myaccount/dashboard.php#L47